controllers/storagecluster/server.go

```diff
@@ -0,0 +1,187 @@
+
+	for _, obj := range []client.Object{
+		desiredServerDeployment,
+		desiredServerService,
+	} {
+		_, err := controllerutil.CreateOrUpdate(
+			context.TODO(), r.Client, obj,
+			func() error { return nil },
+		)
+
+		if err != nil && !errors.IsAlreadyExists(err) {
+			r.Log.Error(err, "nigoyal Failed to create resource", "Kind", obj.GetObjectKind(), "Name", obj.GetName())
+			multierr.AppendInto(&finalErr, err)
+		}
+	}
+
+	if finalErr != nil {
+		return finalErr
+	}
+
+	if err := obj.isDeploymentReady(r, instance); err != nil {
+		r.Log.Error(err, "nigoyal Waiting for deployment to be ready")
+		return err
+	}
+
+	return nil
+}
+
+func (obj *ocsProviderServer) ensureDeleted(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	// We do not check instance.Spec.AllowRemoteStorageConsumers because provider can disable this functionality
+	// and we need to delete the resources even the flag is not enabled (uninstall case).
+
+	// This func is directly called by the ensureCreated if the flag is disabled and deletes the resource
+	// Which means we do not need to call ensureDeleted while reconciling unless we are uninstalling
+
+	// NOTE: Do not add the check
+
+	desiredServerService := GetProviderServerService(instance)
+	desiredServerDeployment := GetProviderServerDeployment(instance)
+
+	var finalErr error
+
+	for _, obj := range []client.Object{
+		desiredServerDeployment,
+		desiredServerService,
+	} {
+		err := r.Client.Delete(context.TODO(), obj)
+
+		if err != nil && !errors.IsNotFound(err) {
+			r.Log.Error(err, "nigoyal Failed to delete resource", "Kind", obj.GetObjectKind(), "Name", obj.GetName())
+			multierr.AppendInto(&finalErr, err)
+		}
+	}
+
+	if finalErr == nil {
+		r.Log.Info("nigoyal Resource deletion for provider succeeded")
+	}
+
+	return finalErr
+}
+
+func (obj *ocsProviderServer) isDeploymentReady(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	desiredServerDeployment := GetProviderServerDeployment(instance)
+	actualServerDeployment := &appsv1.Deployment{}
+
+	err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Name:      desiredServerDeployment.Name,
+		Namespace: desiredServerDeployment.Namespace,
+	}, actualServerDeployment)
+
+	if err != nil {
+		r.Log.Error(err, "nigoyal Failed to get deployment", "Name", desiredServerDeployment.Name)
+		return err
+	}
+
+	if actualServerDeployment.Status.AvailableReplicas != *desiredServerDeployment.Spec.Replicas {
+		r.Log.Error(err, "nigoyal Deployment is not ready", "Name", desiredServerDeployment.Name,
+			"AvailableReplicas", actualServerDeployment.Status.AvailableReplicas)
+		return fmt.Errorf("Deployment %s is not ready", desiredServerDeployment.Name)
+	}
+
+	r.Log.Info("nigoyal Deployment is working as expected", "Name", desiredServerDeployment.Name)
+	return nil
+}
+
+func GetProviderServerDeployment(instance *ocsv1.StorageCluster) *appsv1.Deployment {
+
+	return &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
```

Comment:
Do we want to add owner references and monitor the resources for any changes? same goes for the service also.

Reply:
I'd say yes.

Reply:
done